# Exercise 15

# Fraud Detection

## Introduction

- Fraud Detection Dataset from Microsoft Azure: [data](http://gallery.cortanaintelligence.com/Experiment/8e9fe4e03b8b4c65b9ca947c72b8e463)

Fraud detection is one of the earliest industrial applications of data mining and machine learning. Fraud detection is typically handled as a binary classification problem, but the class population is unbalanced because instances of fraud are usually very rare compared to the overall volume of transactions. Moreover, when fraudulent transactions are discovered, the business typically takes measures to block the accounts from transacting to prevent further losses. 

In [1]:
%matplotlib inline
import pandas as pd
import numpy as np
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn import metrics

In [2]:
import pandas as pd

url = 'https://raw.githubusercontent.com/albahnsen/PracticalMachineLearningClass/master/datasets/15_fraud_detection.csv.zip'
df = pd.read_csv(url, index_col=0)
df.head()

accountAge  digitalItemCount  sumPurchaseCount1Day  sumPurchaseAmount1Day  \
0        2000                 0                     0                   0.00   
1          62                 1                     1                1185.44   
2        2000                 0                     0                   0.00   
3           1                 1                     0                   0.00   
4           1                 1                     0                   0.00   

   sumPurchaseAmount30Day  paymentBillingPostalCode - LogOddsForClass_0  \
0                  720.25                                      5.064533   
1                 2530.37                                      0.538996   
2                    0.00                                      5.064533   
3                    0.00                                      5.064533   
4                  132.73                                      5.412885   

   accountPostalCode - LogOddsForClass_0  \
0                               0.421214   
1                               0.481838   
2                               5.096396   
3                               5.096396   
4                               0.342945   

   paymentBillingState - LogOddsForClass_0  accountState - LogOddsForClass_0  \
0                                 1.312186                          0.566395   
1                                 4.401370                          4.500157   
2                                 3.056357                          3.155226   
3                                 3.331154                          3.331239   
4                                 5.563677                          4.086965   

   paymentInstrumentAgeInAccount  ipState - LogOddsForClass_0  \
0                    3279.574306                     1.218157   
1                      61.970139                     4.035601   
2                       0.000000                     3.314186   
3                       0.000000                     3.529398   
4                       0.001389                     3.529398   

   transactionAmount  transactionAmountUSD  ipPostalCode - LogOddsForClass_0  \
0             599.00            626.164650                          1.259543   
1            1185.44           1185.440000                          3.981118   
2              32.09             32.090000                          5.008490   
3             133.28            132.729554                          1.324925   
4             543.66            543.660000                          2.693451   

   localHour - LogOddsForClass_0  Label  
0                       4.745402      0  
1                       4.921349      0  
2                       4.742303      0  
3                       4.745402      0  
4                       4.876771      0

In [3]:
df.head()

accountAge  digitalItemCount  sumPurchaseCount1Day  sumPurchaseAmount1Day  \
0        2000                 0                     0                   0.00   
1          62                 1                     1                1185.44   
2        2000                 0                     0                   0.00   
3           1                 1                     0                   0.00   
4           1                 1                     0                   0.00   

   sumPurchaseAmount30Day  paymentBillingPostalCode - LogOddsForClass_0  \
0                  720.25                                      5.064533   
1                 2530.37                                      0.538996   
2                    0.00                                      5.064533   
3                    0.00                                      5.064533   
4                  132.73                                      5.412885   

   accountPostalCode - LogOddsForClass_0  \
0                               0.421214   
1                               0.481838   
2                               5.096396   
3                               5.096396   
4                               0.342945   

   paymentBillingState - LogOddsForClass_0  accountState - LogOddsForClass_0  \
0                                 1.312186                          0.566395   
1                                 4.401370                          4.500157   
2                                 3.056357                          3.155226   
3                                 3.331154                          3.331239   
4                                 5.563677                          4.086965   

   paymentInstrumentAgeInAccount  ipState - LogOddsForClass_0  \
0                    3279.574306                     1.218157   
1                      61.970139                     4.035601   
2                       0.000000                     3.314186   
3                       0.000000                     3.529398   
4                       0.001389                     3.529398   

   transactionAmount  transactionAmountUSD  ipPostalCode - LogOddsForClass_0  \
0             599.00            626.164650                          1.259543   
1            1185.44           1185.440000                          3.981118   
2              32.09             32.090000                          5.008490   
3             133.28            132.729554                          1.324925   
4             543.66            543.660000                          2.693451   

   localHour - LogOddsForClass_0  Label  
0                       4.745402      0  
1                       4.921349      0  
2                       4.742303      0  
3                       4.745402      0  
4                       4.876771      0

In [4]:
df.shape, df.Label.sum(), df.Label.mean()

((138721, 16), 797, 0.0057453449730033666)

# Exercise 15.1

Estimate a Logistic Regression and a Decision Tree

Evaluate using the following metrics:
* Accuracy
* F1-Score
* F_Beta-Score (Beta=10)

Comment about the results

In [5]:
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import f1_score
from sklearn.metrics import fbeta_score
from sklearn.metrics import accuracy_score
X = df
y = df.Label
X = X.drop(['Label'], axis=1)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=42)
LR = LogisticRegression()
DT = DecisionTreeClassifier()
RF = RandomForestClassifier()
LR.fit(X_train, y_train)
DT.fit(X_train, y_train)
RF.fit(X_train, y_train)
y_predLR = LR.predict(X_test)
y_predDT = DT.predict(X_test)
y_predRF = RF.predict(X_test)
print("****************** Regresion Logistica: ******************")
print('Accurancy: ', accuracy_score(y_predLR, y_test))
print('F1 Score: ', f1_score(y_predLR, y_test))
print('FBeta Score: ', fbeta_score(y_predLR, y_test, beta=10))
print("****************** Decision Tree: ******************")
print('Accurancy: ', accuracy_score(y_predDT, y_test))
print('F1 Score: ', f1_score(y_predDT, y_test))
print('FBeta Score: ', fbeta_score(y_predDT, y_test, beta=10))
print("****************** Random Forest: ******************")
print('Accurancy: ', accuracy_score(y_predRF, y_test))
print('F1 Score: ', f1_score(y_predRF, y_test))
print('FBeta Score: ', fbeta_score(y_predRF, y_test, beta=10))



C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


****************** Regresion Logistica: ******************
Accurancy:  0.994112982675349
F1 Score:  0.0
FBeta Score:  0.0
****************** Decision Tree: ******************
Accurancy:  0.9888026527620924
F1 Score:  0.13382899628252787
FBeta Score:  0.12306650871551869
****************** Random Forest: ******************
Accurancy:  0.9940649253910662
F1 Score:  0.09523809523809523
FBeta Score:  0.4311986863711002


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1145: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true samples.
  'recall', 'true', average, warn_for)


# Exercise 15.2

Under-sample the negative class using random-under-sampling

Which is parameter for target_percentage did you choose?
How the results change?

**Only apply under-sampling to the training set, evaluate using the whole test set**

In [6]:
def UnderSampling(X, y, target_percentage=0.5, seed=None):
    # Assuming minority class is the positive
    n_samples = y.shape[0]
    n_samples_0 = (y == 0).sum()
    n_samples_1 = (y == 1).sum()

    n_samples_0_new =  n_samples_1 / target_percentage - n_samples_1
    n_samples_0_new_per = n_samples_0_new / n_samples_0

    filter_ = y == 0

    np.random.seed(seed)
    rand_1 = np.random.binomial(n=1, p=n_samples_0_new_per, size=n_samples)
    
    filter_ = filter_ & rand_1
    filter_ = filter_ | (y == 1)
    filter_ = filter_.astype(bool)
    
    return X[filter_], y[filter_]

In [7]:
resultados =[]
for target_percentage in [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]:
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=42)
    X_train, y_train = UnderSampling(X_train, y_train, target_percentage, 1)
    LR.fit(X_train, y_train)
    DT.fit(X_train, y_train)
    RF.fit(X_train, y_train)
    y_predLR = LR.predict(X_test)
    y_predDT = DT.predict(X_test)
    y_predRF = RF.predict(X_test)
    print('Target Percentage: ', target_percentage)
    print("****************** Regresion Logistica: ******************")
    print('Accurancy: ', accuracy_score(y_predLR, y_test))
    print('F1 Score: ', f1_score(y_predLR, y_test))
    print('FBeta Score: ', fbeta_score(y_predLR, y_test, beta=10))
    resultados.append(["RL", "Under-Sampling", accuracy_score(y_predLR, y_test),  f1_score(y_predLR, y_test), fbeta_score(y_predLR, y_test, beta=10)])
    print("****************** Decision Tree: ******************")
    print('Accurancy: ', accuracy_score(y_predDT, y_test))
    print('F1 Score: ', f1_score(y_predDT, y_test))
    print('FBeta Score: ', fbeta_score(y_predDT, y_test, beta=10))
    resultados.append(["DT", "Under-Sampling",  accuracy_score(y_predDT, y_test),  f1_score(y_predDT, y_test), fbeta_score(y_predDT, y_test, beta=10)])
    print("****************** Random Forest: ******************")
    print('Accurancy: ', accuracy_score(y_predRF, y_test))
    print('F1 Score: ', f1_score(y_predRF, y_test))
    print('FBeta Score: ', fbeta_score(y_predRF, y_test, beta=10))
    resultados.append(["RF",  "Under-Sampling", accuracy_score(y_predRF, y_test),  f1_score(y_predRF, y_test), fbeta_score(y_predRF, y_test, beta=10)])

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Target Percentage:  0.1
****************** Regresion Logistica: ******************
Accurancy:  0.9914217747555085
F1 Score:  0.011080332409972301
FBeta Score:  0.017053609117771214
****************** Decision Tree: ******************
Accurancy:  0.9199125357426052
F1 Score:  0.05285592497868712
FBeta Score:  0.02866822322940988
****************** Random Forest: ******************
Accurancy:  0.9827714635845929
F1 Score:  0.15348288075560804
FBeta Score:  0.1086111340888411


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Target Percentage:  0.2
****************** Regresion Logistica: ******************
Accurancy:  0.9779417065141649
F1 Score:  0.04375
FBeta Score:  0.029563035751620323
****************** Decision Tree: ******************
Accurancy:  0.8617151644760555
F1 Score:  0.04226992844067232
FBeta Score:  0.022244188365458814
****************** Random Forest: ******************
Accurancy:  0.9594396520652618
F1 Score:  0.09635974304068523
FBeta Score:  0.05592297517610508


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Target Percentage:  0.3
****************** Regresion Logistica: ******************
Accurancy:  0.950116538914386
F1 Score:  0.04155124653739612
FBeta Score:  0.0236294158933167
****************** Decision Tree: ******************
Accurancy:  0.7869620587740587
F1 Score:  0.0327296530656775
FBeta Score:  0.016977738430763886
****************** Random Forest: ******************
Accurancy:  0.901410481293702
F1 Score:  0.06131320064058568
FBeta Score:  0.032782279063571076
Target Percentage: 

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


 0.4
****************** Regresion Logistica: ******************
Accurancy:  0.8815628228848787
F1 Score:  0.03674027750635138
FBeta Score:  0.01947706920780806
****************** Decision Tree: ******************
Accurancy:  0.721435951654372
F1 Score:  0.024568784181741694
FBeta Score:  0.01266571898526513
****************** Random Forest: ******************
Accurancy:  0.8455919455991542
F1 Score:  0.04118173679498657
FBeta Score:  0.021577688502891115
Target Percentage:  0.5
****************** Regresion Logistica: ******************


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Accurancy:  0.5441766585770238
F1 Score:  0.01933416046319272
FBeta Score:  0.00988773103612553
****************** Decision Tree: ******************
Accurancy:  0.6491337674508013
F1 Score:  0.021838156484458734
FBeta Score:  0.011210415751645336
****************** Random Forest: ******************
Accurancy:  0.7703582670543287
F1 Score:  0.032594392144953944
FBeta Score:  0.016874471408039264
Target Percentage:  0.6
****************** Regresion Logistica: ******************


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Accurancy:  0.3980344570728308
F1 Score:  0.017414496391590838
FBeta Score:  0.00887878685882523
****************** Decision Tree: ******************
Accurancy:  0.6155417257370785
F1 Score:  0.02152641878669276
FBeta Score:  0.011034517007098317
****************** Random Forest: ******************
Accurancy:  0.6739553547829013
F1 Score:  0.026544228423846763
FBeta Score:  0.013642221159497534
Target Percentage:  0.7
****************** Regresion Logistica: ******************
Accurancy:  0.27440709325516016
F1 Score:  0.015261699005380729
FBeta Score:  0.007768605092620746
****************** Decision Tree: ******************
Accurancy:  0.5048898286757816
F1 Score:  0.018014583234046606
FBeta Score:  0.009203754016908986
****************** Random Forest: ******************
Accurancy:  0.5577047841026503
F1 Score:  0.021372747089159448
FBeta Score:  0.010934238909430722
Target Percentage: 

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


 0.8
****************** Regresion Logistica: ******************
Accurancy:  0.12833697767739144
F1 Score:  0.013112791773219435
FBeta Score:  0.006665945510735867
****************** Decision Tree: ******************
Accurancy:  0.3916668669053512
F1 Score:  0.015247578668948618
FBeta Score:  0.007773364563965533
****************** Random Forest: ******************
Accurancy:  0.3517072350241488
F1 Score:  0.016261941223656384
FBeta Score:  0.00828555610123071
Target Percentage:  0.9
****************** Regresion Logistica: ******************
Accurancy:  0.02486964461638273
F1 Score:  0.011930268796260227
FBeta Score:  0.006060576373181613
****************** Decision Tree: ******************
Accurancy:  0.23399091717327053
F1 Score:  0.01410236585743003
FBeta Score:  0.007175519475764549
****************** Random Forest: ******************
Accurancy:  0.11987889564360718
F1 Score:  0.012828805519620524
FBeta Score:  0.006521175916845376


# Exercise 15.3

Same analysis using random-over-sampling

In [8]:
import random
def OverSampling(X, y, target_percentage=0.5, seed=None):
    # Assuming minority class is the positive
    n_samples = y.shape[0]
    n_samples_0 = (y == 0).sum()
    n_samples_1 = (y == 1).sum()

    n_samples_1_new =  -target_percentage * n_samples_0 / (target_percentage- 1)

    np.random.seed(seed)
    filter_ = np.random.choice(X[y == 1].shape[0], int(n_samples_1_new))
    # filter_ is within the positives, change to be of all
    filter_ = np.nonzero(y == 1)[0][filter_]
    
    filter_ = np.concatenate((filter_, np.nonzero(y == 0)[0]), axis=0)
    
    return X[filter_], y[filter_]

In [9]:
for target_percentage in [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]:
    X_t, y_t = OverSampling(X.values, y, target_percentage, 1)
    X_train, X_test, y_train, y_test = train_test_split(X_t, y_t, test_size=0.30, random_state=42)
    LR.fit(X_train, y_train)
    DT.fit(X_train, y_train)
    RF.fit(X_train, y_train)
    y_predLR = LR.predict(X_test)
    y_predDT = DT.predict(X_test)
    y_predRF = RF.predict(X_test)
    print('Target Percentage: ', target_percentage)
    print("****************** Regresion Logistica: ******************")
    print('Accurancy: ', accuracy_score(y_predLR, y_test))
    print('F1 Score: ', f1_score(y_predLR, y_test))
    print('FBeta Score: ', fbeta_score(y_predLR, y_test, beta=10))
    resultados.append(["RL", "Over-Sampling", accuracy_score(y_predLR, y_test),  f1_score(y_predLR, y_test), fbeta_score(y_predLR, y_test, beta=10)])
    print("****************** Decision Tree: ******************")
    print('Accurancy: ', accuracy_score(y_predDT, y_test))
    print('F1 Score: ', f1_score(y_predDT, y_test))
    print('FBeta Score: ', fbeta_score(y_predDT, y_test, beta=10))
    resultados.append(["DT", "Over-Sampling",  accuracy_score(y_predDT, y_test),  f1_score(y_predDT, y_test), fbeta_score(y_predDT, y_test, beta=10)])
    print("****************** Random Forest: ******************")
    print('Accurancy: ', accuracy_score(y_predRF, y_test))
    print('F1 Score: ', f1_score(y_predRF, y_test))
    print('FBeta Score: ', fbeta_score(y_predRF, y_test, beta=10))
    resultados.append(["RF",  "Over-Sampling", accuracy_score(y_predRF, y_test),  f1_score(y_predRF, y_test), fbeta_score(y_predRF, y_test, beta=10)])

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


Target Percentage:  0.1
****************** Regresion Logistica: ******************
Accurancy:  0.9012941816204459
F1 Score:  0.0
FBeta Score:  0.0
****************** Decision Tree: ******************
Accurancy:  0.9928874388254486
F1 Score:  0.9651423089222898
FBeta Score:  0.9343882386840864
****************** Random Forest: ******************
Accurancy:  0.998194671016857
F1 Score:  0.9909160555981176
FBeta Score:  0.9834255688454036


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Target Percentage:  0.2
****************** Regresion Logistica: ******************
Accurancy:  0.8058273075287112
F1 Score:  0.12219211607376976
FBeta Score:  0.5661643824630905
****************** Decision Tree: ******************
Accurancy:  0.9945091063763969
F1 Score:  0.9863971644793562
FBeta Score:  0.9734181207921774
****************** Random Forest: ******************
Accurancy:  0.9985499400641893
F1 Score:  0.9963713774251294
FBeta Score:  0.9928400749596419


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Target Percentage:  0.3
****************** Regresion Logistica: ******************
Accurancy:  0.7362250680922332
F1 Score:  0.33645416631202657
FBeta Score:  0.6576363483626337
****************** Decision Tree: ******************
Accurancy:  0.9950601410904908
F1 Score:  0.9917741844610963
FBeta Score:  0.9838415408529563
****************** Random Forest: ******************
Accurancy:  0.9987819525976552
F1 Score:  0.997959067974375
FBeta Score:  0.9959666192931165


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Target Percentage:  0.4
****************** Regresion Logistica: ******************
Accurancy:  0.6835648618079522
F1 Score:  0.4659585923351769
FBeta Score:  0.6975866832275064
****************** Decision Tree: ******************
Accurancy:  0.9960412981062035
F1 Score:  0.9950342870655001
FBeta Score:  0.9902145346778253
****************** Random Forest: ******************
Accurancy:  0.9990429511905107
F1 Score:  0.9987949607449332
FBeta Score:  0.9976165988721457


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Target Percentage:  0.5
****************** Regresion Logistica: ******************
Accurancy:  0.6543290435623225
F1 Score:  0.6866126205083259
FBeta Score:  0.6257473540488545
****************** Decision Tree: ******************
Accurancy:  0.996531931605341
F1 Score:  0.9965215915839484
FBeta Score:  0.9931354673658443
****************** Random Forest: ******************
Accurancy:  0.9993112198658691
F1 Score:  0.9993072351389783
FBeta Score:  0.998629119286103


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Target Percentage:  0.6
****************** Regresion Logistica: ******************
Accurancy:  0.685623966822308
F1 Score:  0.7721604125213687
FBeta Score:  0.6842926620796391
****************** Decision Tree: ******************
Accurancy:  0.9974285355219783
F1 Score:  0.9978600160901046
FBeta Score:  0.9957712782996468
****************** Random Forest: ******************
Accurancy:  0.9993909689394159
F1 Score:  0.9994923326107803
FBeta Score:  0.998995218035294


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Target Percentage:  0.7
****************** Regresion Logistica: ******************
Accurancy:  0.7358400278414199
F1 Score:  0.8340348383805254
FBeta Score:  0.7452371102046962
****************** Decision Tree: ******************
Accurancy:  0.9980133986833328
F1 Score:  0.998581163652934
FBeta Score:  0.9971943250411303
****************** Random Forest: ******************
Accurancy:  0.9996012296627128
F1 Score:  0.9997148737409084
FBeta Score:  0.9994355512985861


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Target Percentage:  0.8
****************** Regresion Logistica: ******************
Accurancy:  0.8018328934775674
F1 Score:  0.8861836918259247
FBeta Score:  0.8204104773379918
****************** Decision Tree: ******************
Accurancy:  0.99877227071914
F1 Score:  0.9992326376720523
FBeta Score:  0.9984816127185857
****************** Random Forest: ******************
Accurancy:  0.9995988128727898
F1 Score:  0.9997491181351316
FBeta Score:  0.9995033263669371


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


Target Percentage:  0.9
****************** Regresion Logistica: ******************
Accurancy:  0.9011605425210019
F1 Score:  0.9479052953247623
FBeta Score:  0.9028555534979446
****************** Decision Tree: ******************
Accurancy:  0.9994078864688766
F1 Score:  0.9996712878673023
FBeta Score:  0.9993492945449848
****************** Random Forest: ******************
Accurancy:  0.9998356582852392
F1 Score:  0.999908743940866
FBeta Score:  0.9998193110948516


# Exercise 15.4 (3 points)

Evaluate the results using SMOTE

Which parameters did you choose?

In [10]:
def SMOTE(X, y, target_percentage=0.5, k=5, seed=None):
    # Calculate the NearestNeighbors
    from sklearn.neighbors import NearestNeighbors
    nearest_neighbour_ = NearestNeighbors(n_neighbors=k + 1)
    nearest_neighbour_.fit(X[y==1])
    nns = nearest_neighbour_.kneighbors(X[y==1], 
                                    return_distance=False)[:, 1:]
    
    # New samples
    n_samples_1_new =  int(-target_percentage * n_samples_0 / (target_percentage- 1) - n_samples_1)
    
    # A matrix to store the synthetic samples
    new = np.zeros((n_samples_1_new, X.shape[1]))
    
    # Create seeds
    np.random.seed(seed)
    seeds = np.random.randint(1, 1000000, 3)
    
    # Select examples to use as base
    np.random.seed(seeds[0])
    sel_ = np.random.choice(y[y==1].shape[0], n_samples_1_new)
    
    # Define random seeds (2 per example)
    np.random.seed(seeds[1])
    nn__=[]
    # Select one random neighbor for each example to use as base
    for i, sel in enumerate(sel_):
        nn__.append(np.random.choice(nns[sel]))
    
    np.random.seed(seeds[2])
    steps = np.random.uniform(size=n_samples_1_new)  

    # For each selected examples create one synthetic case
    for i, sel in enumerate(sel_):
        # Select neighbor
        nn_ = nn__[i]
        step = steps[i]
        # Create new sample
        new[i, :] = X[y==1][sel] - step * (X[y==1][sel] - X[y==1][nn_])
    
    X = np.vstack((X, new))
    y = np.append(y, np.ones(n_samples_1_new))
    
    return X, y

In [11]:
for target_percentage in [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]:
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=42)
    X_train, y_train = UnderSampling(X_train, y_train, target_percentage, 1)
    LR.fit(X_train, y_train)
    DT.fit(X_train, y_train)
    RF.fit(X_train, y_train)
    y_predLR = LR.predict(X_test)
    y_predDT = DT.predict(X_test)
    y_predRF = RF.predict(X_test)
    print('Target Percentage: ', target_percentage)
    print("****************** Regresion Logistica: ******************")
    print('Accurancy: ', accuracy_score(y_predLR, y_test))
    print('F1 Score: ', f1_score(y_predLR, y_test))
    print('FBeta Score: ', fbeta_score(y_predLR, y_test, beta=10))
    resultados.append(["RL", "SMOTE-Sampling", accuracy_score(y_predLR, y_test),  f1_score(y_predLR, y_test), fbeta_score(y_predLR, y_test, beta=10)])
    print("****************** Decision Tree: ******************")
    print('Accurancy: ', accuracy_score(y_predDT, y_test))
    print('F1 Score: ', f1_score(y_predDT, y_test))
    print('FBeta Score: ', fbeta_score(y_predDT, y_test, beta=10))
    resultados.append(["DT", "SMOTE-Sampling",  accuracy_score(y_predDT, y_test),  f1_score(y_predDT, y_test), fbeta_score(y_predDT, y_test, beta=10)])
    print("****************** Random Forest: ******************")
    print('Accurancy: ', accuracy_score(y_predRF, y_test))
    print('F1 Score: ', f1_score(y_predRF, y_test))
    print('FBeta Score: ', fbeta_score(y_predRF, y_test, beta=10))
    resultados.append(["RF",  "SMOTE-Sampling", accuracy_score(y_predRF, y_test),  f1_score(y_predRF, y_test), fbeta_score(y_predRF, y_test, beta=10)])

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Target Percentage:  0.1
****************** Regresion Logistica: ******************
Accurancy:  0.9914217747555085
F1 Score:  0.011080332409972301
FBeta Score:  0.017053609117771214
****************** Decision Tree: ******************
Accurancy:  0.9199125357426052
F1 Score:  0.05285592497868712
FBeta Score:  0.02866822322940988
****************** Random Forest: ******************
Accurancy:  0.9827714635845929
F1 Score:  0.15348288075560804
FBeta Score:  0.1086111340888411


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Target Percentage:  0.2
****************** Regresion Logistica: ******************
Accurancy:  0.9779417065141649
F1 Score:  0.04375
FBeta Score:  0.029563035751620323
****************** Decision Tree: ******************
Accurancy:  0.8617151644760555
F1 Score:  0.04226992844067232
FBeta Score:  0.022244188365458814
****************** Random Forest: ******************
Accurancy:  0.9594396520652618
F1 Score:  0.09635974304068523
FBeta Score:  0.05592297517610508


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Target Percentage:  0.3
****************** Regresion Logistica: ******************
Accurancy:  0.950116538914386
F1 Score:  0.04155124653739612
FBeta Score:  0.0236294158933167
****************** Decision Tree: ******************
Accurancy:  0.7869620587740587
F1 Score:  0.0327296530656775
FBeta Score:  0.016977738430763886
****************** Random Forest: ******************
Accurancy:  0.901410481293702
F1 Score:  0.06131320064058568
FBeta Score:  0.032782279063571076


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Target Percentage:  0.4
****************** Regresion Logistica: ******************
Accurancy:  0.8815628228848787
F1 Score:  0.03674027750635138
FBeta Score:  0.01947706920780806
****************** Decision Tree: ******************
Accurancy:  0.721435951654372
F1 Score:  0.024568784181741694
FBeta Score:  0.01266571898526513
****************** Random Forest: ******************
Accurancy:  0.8455919455991542
F1 Score:  0.04118173679498657
FBeta Score:  0.021577688502891115
Target Percentage:  0.5
****************** Regresion Logistica: ******************
Accurancy:  0.5441766585770238


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


F1 Score:  0.01933416046319272
FBeta Score:  0.00988773103612553
****************** Decision Tree: ******************
Accurancy:  0.6491337674508013
F1 Score:  0.021838156484458734
FBeta Score:  0.011210415751645336
****************** Random Forest: ******************
Accurancy:  0.7703582670543287
F1 Score:  0.032594392144953944
FBeta Score:  0.016874471408039264
Target Percentage:  0.6
****************** Regresion Logistica: ******************
Accurancy:  0.3980344570728308
F1 Score:  0.017414496391590838
FBeta Score: 

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


 0.00887878685882523
****************** Decision Tree: ******************
Accurancy:  0.6155417257370785
F1 Score:  0.02152641878669276
FBeta Score:  0.011034517007098317
****************** Random Forest: ******************
Accurancy:  0.6739553547829013
F1 Score:  0.026544228423846763
FBeta Score:  0.013642221159497534
Target Percentage:  0.7
****************** Regresion Logistica: ******************
Accurancy:  0.27440709325516016
F1 Score:  0.015261699005380729
FBeta Score:  0.007768605092620746
****************** Decision Tree: ******************

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)



Accurancy:  0.5048898286757816
F1 Score:  0.018014583234046606
FBeta Score:  0.009203754016908986
****************** Random Forest: ******************
Accurancy:  0.5577047841026503
F1 Score:  0.021372747089159448
FBeta Score:  0.010934238909430722
Target Percentage:  0.8
****************** Regresion Logistica: ******************
Accurancy:  0.12833697767739144
F1 Score:  0.013112791773219435
FBeta Score:  0.006665945510735867
****************** Decision Tree: ******************


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Accurancy:  0.3916668669053512
F1 Score:  0.015247578668948618
FBeta Score:  0.007773364563965533
****************** Random Forest: ******************
Accurancy:  0.3517072350241488
F1 Score:  0.016261941223656384
FBeta Score:  0.00828555610123071
Target Percentage:  0.9
****************** Regresion Logistica: ******************
Accurancy:  0.02486964461638273
F1 Score:  0.011930268796260227
FBeta Score:  0.006060576373181613
****************** Decision Tree: ******************
Accurancy:  0.23399091717327053
F1 Score:  0.01410236585743003
FBeta Score:  0.007175519475764549
****************** Random Forest: ******************
Accurancy:  0.11987889564360718
F1 Score:  0.012828805519620524
FBeta Score:  0.006521175916845376


# Exercise 15.5 (3 points)

Evaluate the results using Adaptive Synthetic Sampling Approach for Imbalanced
Learning (ADASYN)

http://www.ele.uri.edu/faculty/he/PDFfiles/adasyn.pdf
https://imbalanced-learn.readthedocs.io/en/stable/generated/imblearn.over_sampling.ADASYN.html#rf9172e970ca5-1

In [12]:
from collections import Counter
from imblearn.over_sampling import ADASYN 
ada = ADASYN(random_state=42)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=42)
X_train, y_train = ada.fit_resample(X_train, y_train)
LR.fit(X_train, y_train)
DT.fit(X_train, y_train)
RF.fit(X_train, y_train)
y_predLR = LR.predict(X_test)
y_predDT = DT.predict(X_test)
y_predRF = RF.predict(X_test)
print("****************** Regresion Logistica: ******************")
print('Accurancy: ', accuracy_score(y_predLR, y_test))
print('F1 Score: ', f1_score(y_predLR, y_test))
print('FBeta Score: ', fbeta_score(y_predLR, y_test, beta=10))
resultados.append(["RL", "ADASYN-Sampling", accuracy_score(y_predLR, y_test),  f1_score(y_predLR, y_test), fbeta_score(y_predLR, y_test, beta=10)])
print("****************** Decision Tree: ******************")
print('Accurancy: ', accuracy_score(y_predDT, y_test))
print('F1 Score: ', f1_score(y_predDT, y_test))
print('FBeta Score: ', fbeta_score(y_predDT, y_test, beta=10))
resultados.append(["DT", "ADASYN-Sampling",  accuracy_score(y_predDT, y_test),  f1_score(y_predDT, y_test), fbeta_score(y_predDT, y_test, beta=10)])
print("****************** Random Forest: ******************")
print('Accurancy: ', accuracy_score(y_predRF, y_test))
print('F1 Score: ', f1_score(y_predRF, y_test))
print('FBeta Score: ', fbeta_score(y_predRF, y_test, beta=10))
resultados.append(["RF",  "ADASYN-Sampling", accuracy_score(y_predRF, y_test),  f1_score(y_predRF, y_test), fbeta_score(y_predRF, y_test, beta=10)])

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


****************** Regresion Logistica: ******************
Accurancy:  0.5725064276617728
F1 Score:  0.01907702486629542
FBeta Score:  0.009764480048283125
****************** Decision Tree: ******************
Accurancy:  0.9835163514909773
F1 Score:  0.08288770053475936
FBeta Score:  0.06194480166188545
****************** Random Forest: ******************
Accurancy:  0.9925270922940145
F1 Score:  0.15258855585831063
FBeta Score:  0.2272398553635998


# Exercise 15.6 (3 points)

Compare and comment about the results

In [13]:
resultados = pd.DataFrame(resultados, columns=["Model", "Method", "Acc", "F1 Score", "FBeta Score-b10"])

In [14]:
resultados

Model           Method       Acc  F1 Score  FBeta Score-b10
0     RL   Under-Sampling  0.991422  0.011080         0.017054
1     DT   Under-Sampling  0.919913  0.052856         0.028668
2     RF   Under-Sampling  0.982771  0.153483         0.108611
3     RL   Under-Sampling  0.977942  0.043750         0.029563
4     DT   Under-Sampling  0.861715  0.042270         0.022244
5     RF   Under-Sampling  0.959440  0.096360         0.055923
6     RL   Under-Sampling  0.950117  0.041551         0.023629
7     DT   Under-Sampling  0.786962  0.032730         0.016978
8     RF   Under-Sampling  0.901410  0.061313         0.032782
9     RL   Under-Sampling  0.881563  0.036740         0.019477
10    DT   Under-Sampling  0.721436  0.024569         0.012666
11    RF   Under-Sampling  0.845592  0.041182         0.021578
12    RL   Under-Sampling  0.544177  0.019334         0.009888
13    DT   Under-Sampling  0.649134  0.021838         0.011210
14    RF   Under-Sampling  0.770358  0.032594         0.016874
15    RL   Under-Sampling  0.398034  0.017414         0.008879
16    DT   Under-Sampling  0.615542  0.021526         0.011035
17    RF   Under-Sampling  0.673955  0.026544         0.013642
18    RL   Under-Sampling  0.274407  0.015262         0.007769
19    DT   Under-Sampling  0.504890  0.018015         0.009204
20    RF   Under-Sampling  0.557705  0.021373         0.010934
21    RL   Under-Sampling  0.128337  0.013113         0.006666
22    DT   Under-Sampling  0.391667  0.015248         0.007773
23    RF   Under-Sampling  0.351707  0.016262         0.008286
24    RL   Under-Sampling  0.024870  0.011930         0.006061
25    DT   Under-Sampling  0.233991  0.014102         0.007176
26    RF   Under-Sampling  0.119879  0.012829         0.006521
27    RL    Over-Sampling  0.901294  0.000000         0.000000
28    DT    Over-Sampling  0.992887  0.965142         0.934388
29    RF    Over-Sampling  0.998195  0.990916         0.983426
..   ...              ...       ...       ...              ...
54    RL   SMOTE-Sampling  0.991422  0.011080         0.017054
55    DT   SMOTE-Sampling  0.919913  0.052856         0.028668
56    RF   SMOTE-Sampling  0.982771  0.153483         0.108611
57    RL   SMOTE-Sampling  0.977942  0.043750         0.029563
58    DT   SMOTE-Sampling  0.861715  0.042270         0.022244
59    RF   SMOTE-Sampling  0.959440  0.096360         0.055923
60    RL   SMOTE-Sampling  0.950117  0.041551         0.023629
61    DT   SMOTE-Sampling  0.786962  0.032730         0.016978
62    RF   SMOTE-Sampling  0.901410  0.061313         0.032782
63    RL   SMOTE-Sampling  0.881563  0.036740         0.019477
64    DT   SMOTE-Sampling  0.721436  0.024569         0.012666
65    RF   SMOTE-Sampling  0.845592  0.041182         0.021578
66    RL   SMOTE-Sampling  0.544177  0.019334         0.009888
67    DT   SMOTE-Sampling  0.649134  0.021838         0.011210
68    RF   SMOTE-Sampling  0.770358  0.032594         0.016874
69    RL   SMOTE-Sampling  0.398034  0.017414         0.008879
70    DT   SMOTE-Sampling  0.615542  0.021526         0.011035
71    RF   SMOTE-Sampling  0.673955  0.026544         0.013642
72    RL   SMOTE-Sampling  0.274407  0.015262         0.007769
73    DT   SMOTE-Sampling  0.504890  0.018015         0.009204
74    RF   SMOTE-Sampling  0.557705  0.021373         0.010934
75    RL   SMOTE-Sampling  0.128337  0.013113         0.006666
76    DT   SMOTE-Sampling  0.391667  0.015248         0.007773
77    RF   SMOTE-Sampling  0.351707  0.016262         0.008286
78    RL   SMOTE-Sampling  0.024870  0.011930         0.006061
79    DT   SMOTE-Sampling  0.233991  0.014102         0.007176
80    RF   SMOTE-Sampling  0.119879  0.012829         0.006521
81    RL  ADASYN-Sampling  0.572506  0.019077         0.009764
82    DT  ADASYN-Sampling  0.983516  0.082888         0.061945
83    RF  ADASYN-Sampling  0.992527  0.152589         0.227240

[84 rows x 5 columns]

In [17]:
resultados.describe()

Acc   F1 Score  FBeta Score-b10
count  84.000000  84.000000        84.000000
mean    0.731584   0.298203         0.296346
std     0.285118   0.414934         0.418476
min     0.024870   0.000000         0.000000
25%     0.557705   0.018811         0.009888
50%     0.825710   0.041366         0.021911
75%     0.985493   0.787629         0.709499
max     0.999836   0.999909         0.999819

Se puede evidenciar que la media de los modelos calibrados bajo diferentes metodologías es de 0.73 y tanto el F1 Score como el FBeta Score a b=10 no dan valores superior a 1 sino tendenciales a 0. Adicionalmente el percentil 50% muestra que los modelos evaluados tenian un Acc de .82 y aquellos que lograban tener un Acc de .98 lograban casi un F1 y un Beta Score de .78